### DISCUSSION TITLE
How to initialize edge feature/edge index tensors in Heterogeneous Graphs for edges with no features?

### DISCUSSION THREAD

**Discussion by stefanschutz** [pyg-team/pytorch_geometric#8879] [Q&A] [answered]
Hi everyone, I have some questions regarding creating heterogeneous graphs. In the case that I have only one node type and multiple edge types, and only 1 edge type has edge features, should I initialize the empty tensors for those edge types that don't have edge features? For example:

```python
data = HeteroData()
data["node"].x = ... # [num_nodes, num_node_feats], only 1 type

data["node", "type1", "node"].edge_index = ... #[2, num_edge_type1]
data["node", "type2", "node"].edge_index = ... #[2, num_edge_type2]

data["node", "type1", "node"].edge_attr = ... #[num_edge_type1, num_edge_type1_feat]
data["node", "type2", "node"].edge_attr = ... # zeros-tensor or no init?
```

Another question popping up is some graphs might not have a specific edge type, which leads to the `edge_index` of that type having empty tensor. In this case, using `utils.to_dgl` to convert pyg object to dgl will have an error since it assumes `edge_index` of all edge types are not empty. What could be a good solution for handling this case for learning on GNNs?

```python
data["node", "type3", "node"].edge_index = ... # [0]
dgl_data = to_dgl(data) # ValueError: not enough values to unpack (expected 2, got 0)
```

Thank you in advance!

**Comment by rusty1s** [comment]
In general, node types/ edge types / their features need to be consistent across graphs. If an edge type of all graphs does not have edge features, then you don't need to specify them.

**Reply by stefanschutz** [reply]
Thanks for the quick reply! And sorry for my sudden update, could you also answer the second question about the index as well? I believe this would cause some inconsistency for the edge index if converting to dgl.

**Reply by rusty1s** [reply, marked as answer]
The answer would be the same, right? `edge_index` would need to be defined for the edge types that exist in the union of graphs.
```python
data[edge_type].edge_index = torch.empty(2, 0, dtype=torch.long)
```
Maybe I am misunderstanding the question.

**Reply by stefanschutz** [reply]
Initializing with the size `torch.empty(2, 0, dtype=torch.long)` works for me. I left the empty `edge_index` with long `tensor([0])`, which causes the mismatched size when converting. Thank you!
